File: misc/update-headers.pl
#!/usr/bin/perl -w
#                               POK header
# 
# The following file is a part of the POK project. Any modification should
# be made according to the POK licence. You CANNOT use this file or a part 
# of a file for your own project.
#
# For more information on the POK licence, please see our LICENCE FILE
#
# Please follow the coding guidelines described in doc/CODING_GUIDELINES
#
#                                      Copyright (c) 2007-2009 POK team 
#
# Created by julien on Sun Aug 10 12:45:15 2008 
#


#!/usr/bin/perl -w

use strict;

my $user;
my $date;
my $force;
my $year;

#####################################################
# COMMON FUNCTIONS
#####################################################
sub has_header 
{
	my $filename = shift;
	my $c;
	my $i = 3;
	open (MYFILE, $filename);
	
	while ($i > 0)
	{
		$c = <MYFILE>;
		if ((defined ($c)) && ($c =~ m/.*POK header.*/))
		{
			close (MYFILE);
			return 1;
		}
		$i--;
	}
	close (MYFILE);
	return 0;
}

sub get_file_content
{
	my $filename = shift;
	my @c;
	my $t;

	return if ! -f $filename;

	open (MYFILE, $filename);

	while ($t = <MYFILE>)
	{
		push (@c,$t);
	}

	close (MYFILE);

	return @c;
}


sub treat_file
{
	my $filename = shift;
	treat_c_file ($filename) if ($filename =~ m/\.c$/);
	treat_c_file ($filename) if ($filename =~ m/\.h$/);
	treat_latex_file ($filename) if ($filename =~ m/\.tex$/);
	treat_txt_file ($filename) if ($filename =~ m/\.txt$/);
	treat_perl_file ($filename) if ($filename =~ m/\.pl$/);
	treat_perl_file ($filename) if ($filename =~ m/\.pm$/);
	treat_aadl_file ($filename) if ($filename =~ m/\.aadl$/);
}

sub treat_dir 
{
	my $dirname = shift;
	my $e;
	my $cname;
	my @elts;

	opendir (MYDIR, $dirname) or die ("cannot open rootdir");

	@elts = readdir (MYDIR);

	closedir (MYDIR);

	foreach $e (@elts)
	{
		next if ($e eq ".");
		next if ($e eq "..");
      next if ($e =~ m/^\./);

		
		$cname = "$dirname/$e";

		treat_file ($cname) if -f $cname;
		treat_dir ($cname) if -d $cname;
	}

}




#####################################################
# C FUNCTIONS
#####################################################
sub add_c_header
{
	my $filename = shift;
	my @contents;
	my $v;
	return if ! -f $filename;

	@contents = get_file_content ($filename);
	open (MYFILE, ">$filename");
	seek (MYFILE,0,0);
	print MYFILE "/*\n";
	print MYFILE " *                               POK header\n";
	print MYFILE " * \n";
	print MYFILE " * The following file is a part of the POK project. Any modification should\n";
	print MYFILE " * made according to the POK licence. You CANNOT use this file or a part of\n";
	print MYFILE " * this file is this part of a file for your own project\n";
	print MYFILE " *\n";
	print MYFILE " * For more information on the POK licence, please see our LICENCE FILE\n";
	print MYFILE " *\n";
	print MYFILE " * Please follow the coding guidelines described in doc/CODING_GUIDELINES\n";
	print MYFILE " *\n";
	print MYFILE " *                                      Copyright (c) 2007-$year POK team \n";
	print MYFILE " *\n";
	print MYFILE " * Created by $user on $date \n";
	print MYFILE " */\n\n";

	foreach $v (@contents)
	{
		print MYFILE $v;
	}
	close (MYFILE);
}


sub remove_c_header
{
	my $filename = shift;
	my @contents;
	my $v;
	my $i;
	return if ! -f $filename;

	@contents = get_file_content ($filename);

	$i = 16;

	unlink ($filename) or die ("cannot remove $filename");
	open (MYFILE, ">$filename");
	seek (MYFILE,0,0);
	
	foreach $v (@contents)
	{
		if ($i == 0)
		{
			print MYFILE $v;
		}
		else
		{
			$i--;
		}
	}
	close (MYFILE);
}

sub treat_c_file
{
	my $filename = shift;


	if ((has_header ($filename) == 1) && ($force == 1))
	{
		remove_c_header ($filename);
	}

	if (has_header ($filename) == 0)
	{
		add_c_header ($filename);
	}
}

#####################################################
# LaTeX FUNCTIONS
#####################################################

sub add_latex_header
{
	my $filename = shift;
	my @contents;
	my $v;
	return if ! -f $filename;

	@contents = get_file_content ($filename);
	open (MYFILE, ">$filename");
	seek (MYFILE,0,0);
	print MYFILE "%\n";
	print MYFILE "%                              POK header\n";
	print MYFILE "% \n";
	print MYFILE "% The following file is a part of the POK project. Any modification should\n";
	print MYFILE "% be made according to the POK licence. You CANNOT use this file or a part\n";
	print MYFILE "% of a file for your own project.\n";
	print MYFILE "%\n";
	print MYFILE "% For more information on the POK licence, please see our LICENCE FILE\n";
	print MYFILE "%\n";
	print MYFILE "% Please follow the coding guidelines described in doc/CODING_GUIDELINES\n";
	print MYFILE "%\n";
	print MYFILE "%                                      Copyright (c) 2007-$year POK team \n";
	print MYFILE "%\n";
	print MYFILE "% Created by $user on $date \n";
	print MYFILE "%\n\n\n";

	foreach $v (@contents)
	{
		print MYFILE $v;
	}

	close (MYFILE);
}


sub treat_latex_file
{
	my $filename = shift;
	if (has_header ($filename) == 0)
	{
		add_latex_header ($filename);
	}
}


#####################################################
# TEXT FUNCTIONS
#####################################################
sub treat_txt_file
{
	my $filename = shift;
	# Nothing to add at this moment
}


#####################################################
# PERL FUNCTIONS
#####################################################
sub add_perl_header
{
	my $filename = shift;
	my @contents;
	my $v;
	return if ! -f $filename;

	@contents = get_file_content ($filename);
	open (MYFILE, ">$filename");
	seek (MYFILE,0,0);
	print MYFILE $contents[0];
	print MYFILE "#                               POK header\n";
	print MYFILE "# \n";
	print MYFILE "# The following file is a part of the POK project. Any modification should\n";
	print MYFILE "# be made according to the POK licence. You CANNOT use this file or a part \n";
	print MYFILE "# of a file for your own project.\n";
	print MYFILE "#\n";
	print MYFILE "# For more information on the POK licence, please see our LICENCE FILE\n";
	print MYFILE "#\n";
	print MYFILE "# Please follow the coding guidelines described in doc/CODING_GUIDELINES\n";
	print MYFILE "#\n";
	print MYFILE "#                                      Copyright (c) 2007-$year POK team \n";
	print MYFILE "#\n";
	print MYFILE "# Created by $user on $date \n";
	print MYFILE "#\n\n\n";

	foreach $v (@contents)
	{
		print MYFILE $v;
	}
	close (MYFILE);
}

sub treat_perl_file
{
	my $filename = shift;
	if (has_header ($filename) == 0)
	{
		add_perl_header ($filename);
	}
}



#####################################################
# AADL FUNCTIONS
#####################################################
sub add_aadl_header
{
	my $filename = shift;
	my @contents;
	my $v;
	return if ! -f $filename;

	@contents = get_file_content ($filename);

	open (MYFILE, ">$filename");
	seek (MYFILE,0,0);
	print MYFILE "--\n";
	print MYFILE "--                              POK header\n";
	print MYFILE "-- \n";
	print MYFILE "--  The following file is a part of the POK project. Any modification should\n";
	print MYFILE "--  be made according to the POK licence. You CANNOT use this file or a part \n";
	print MYFILE "--  of a file for your own project.\n";
	print MYFILE "-- \n";
	print MYFILE "--  For more information on the POK licence, please see our LICENCE FILE\n";
	print MYFILE "--\n";
	print MYFILE "--  Please follow the coding guidelines described in doc/CODING_GUIDELINES\n";
	print MYFILE "--\n";
	print MYFILE "--                                      Copyright (c) 2007-$year POK team \n";
	print MYFILE "--\n";
	print MYFILE "--  Created by $user on $date \n";
	print MYFILE "--\n";

	foreach $v (@contents)
	{
		print MYFILE $v;
	}
	close (MYFILE);
}

sub remove_aadl_header
{
	my $filename = shift;
	my @contents;
	my $v;
	my $i;
	return if ! -f $filename;

	@contents = get_file_content ($filename);

	$i = 16;

	unlink ($filename) or die ("cannot remove $filename");
	open (MYFILE, ">$filename");
	seek (MYFILE,0,0);
	
	foreach $v (@contents)
	{
		if ($i == 0)
		{
			print MYFILE $v;
		}
		else
		{
			$i--;
		}
	}
	close (MYFILE);
}



sub treat_aadl_file
{
	my $filename = shift;

	if ((has_header ($filename) == 1) && ($force == 1))
	{
		remove_aadl_header ($filename);
	}

	if (has_header ($filename) == 0)
	{
		add_aadl_header ($filename);
	}

}


#####################################################
# MAIN ENTRYPOINT
#####################################################

$user = $ENV{'USER'};
$date = gmtime();
my @date_array = gmtime();
$year = $date_array[5] + 1900;
$force = 0;
$force = 1 if (defined ($ARGV[0]) && ($ARGV[0] eq '--force'));
treat_dir ($ENV{'PWD'});

File: misc/update-headers.py
#! /usr/bin/python
#
# Usage: update-headers.py [--force]
#
#                               POK header
#
# The following file is a part of the POK project. Any modification should
# be made according to the POK licence. You CANNOT use this file or a part
# of a file for your own project.
#
# For more information on the POK licence, please see our LICENCE FILE
#
# Please follow the coding guidelines described in doc/CODING_GUIDELINES
#
#                                      Copyright (c) 2007-2020 POK team

from datetime import datetime
import itertools
import os
import sys

MARKER = "POK header"

HEADER = """
                              {marker}

The following file is a part of the POK project. Any modification should
be made according to the POK licence. You CANNOT use this file or a part
of a file for your own project.

For more information on the POK licence, please see our LICENCE FILE

Please follow the coding guidelines described in doc/CODING_GUIDELINES

                                     Copyright (c) 2007-{year} POK team
""".format(marker = MARKER, year = datetime.now().year).split("\n")[1:-1]

extensions = {}

def register(handler, *exts):
    """Register a handler for a list of suffixes."""
    for ext in exts:
        extensions[ext] = handler

class Lang(object):

    def __init__(self, begin_delimeter, middle_delimeter = None, end_delimeter = "", empty_after = 1):
        self.begin_delimeter = begin_delimeter if begin_delimeter != middle_delimeter else ""
        self.middle_delimeter = middle_delimeter or begin_delimeter
        self.end_delimeter = end_delimeter
        self.empty_after = empty_after

    def has_header(self, content):
        """Return True if the header marker is present."""
        return any(MARKER in l for l in content)

    def fix_header(self, filename, force):
        """Fix a header in place."""
        try:
            content = [l.strip("\n") for l in open(filename)]
            content = self.add_header(content, force)
            open(filename, "w").write("\n".join(content) + "\n")
        except Exception as e:
            print("Error while handling file {}: {}".format(filename, e))
            raise

    def add_header(self, content, force):
        """Add a header if absent, or after removing an existing one if force is True."""
        if self.has_header(content):
            if force: (before, content) = self.remove_header(content)
            else: return content
        else:
            before = []
        return before + \
            [l.rstrip() for l in ([self.begin_delimeter] if self.begin_delimeter != self.middle_delimeter else []) + \
             ["{} {}".format(self.middle_delimeter, l) for l in HEADER]] + \
             ([self.end_delimeter] if self.end_delimeter else []) + \
             ([""] * self.empty_after + content if content else [])

    def remove_header(self, content):
        """Remove the header, preserving the lines before and after it. The lines before
        the header are returned as the first element of the couple, the rest of the file
        without the header as the second element."""
        i = 0
        (bd, md, ed) = self.begin_delimeter.strip(), self.middle_delimeter.strip(), self.end_delimeter.strip()
        begin_found = 0
        while MARKER not in content[i]:
            if content[i].strip().startswith(bd):
                begin_found = i
            i += 1
        if self.begin_delimeter != self.middle_delimeter:
            i = begin_found
        while i < len(content):
            l = content[i].strip()
            if (bd and l.startswith(bd)) or l.startswith(md):
                del content[i]
                continue
            if ed and l.startswith(ed):
                del content[i]
                break
            break
        return content[:i], list(itertools.dropwhile(lambda s: s.strip() == "", content[i:]))

register(Lang("--"), "ads", "adb", "aadl")  # Ada and AADL
register(Lang("/*", " *", " */"), "c", "h")  # C
register(Lang("%"), "tex")                  # LaTeX
register(Lang("#"), "pm", "pl", "py")       # Perl and Python

force = sys.argv[1:2] == ["--force"]

for filename in (f.strip() for f in os.popen("git ls-files")):
    handler = extensions.get(filename.split(".")[-1])
    if handler:
        handler.fix_header(filename, force)
